File: 백준/Silver/1926. 그림/그림.py
import sys
from collections import deque
input = sys.stdin.readline

n, m = map(int, input().split())
g = [list(map(int, input().split())) for _ in range(n)]
v = [[0]*m for _ in range(n)]
d = [(0, 1), (0, -1), (1, 0), (-1, 0)]
max_res = 0
draw = 0

def in_range(y, x):
    if 0 <= y < n and 0<= x < m:
        return True

def bfs(i, j):
    global v, max_res
    q = deque([(i,j)])
    v[i][j] = 1
    cnt = 1

    while q:
        a, b = q.popleft()
        for dy, dx in d:
            ny, nx = a + dy, b + dx

            if not in_range(ny, nx):
                continue
            if v[ny][nx] == 1:
                continue
            if g[ny][nx] == 1:
                q.append((ny, nx))
                cnt += 1
                v[ny][nx] = 1
    
    max_res = max(max_res, cnt)

for i in range(n):
    for j in range(m):
        if g[i][j] == 1 and v[i][j] == 0:
            bfs(i, j)
            draw += 1

print(draw)
print(max_res)

File: 백준/Silver/1926. 그림/그림.js
const [first, ...input] = require('fs').readFileSync(process.platform == 'win32' ? "./input.txt" : '/dev/stdin').toString().trim().split("\n")

const [m, n] = first.split(" ").map(Number);
const graph = input.map((line) => {
    return line.trim().split(" ").map(Number);
})
//console.log(JSON.stringify(graph))
const dx = [0, 0, -1, 1];
const dy = [-1, 1, 0, 0];
const res = [];
const arr = [];
const visited = Array(m).fill(null).map(() => Array(n).fill(0));

const solution = () => {
    const dfs = (j, k) => {
        let cnt = 1;
        arr.push([j, k]);

        while (arr.length > 0) {
            const [nx, ny] = arr.pop();
            for (let i = 0; i < 4; i++){
                if (0 <= dx[i] + nx && nx + dx[i] < m && 0 <= ny + dy[i] && ny + dy[i] < n
                    && visited[nx + dx[i]][ny + dy[i]] === 0
                    && graph[nx + dx[i]][ny + dy[i]] === 1) {
                    cnt += 1;
                    visited[nx + dx[i]][ny + dy[i]] = 1;
                    arr.push([nx + dx[i], ny + dy[i]]);
                }
            }
        }
        res.push(cnt);
    }

    for (let i = 0; i < m; i++){
        for (let j = 0; j < n; j++){
            if (visited[i][j] === 0 && graph[i][j] === 1) { 
                visited[i][j] = 1;
                dfs(i, j);
            }
        }
    }

    console.log(res.length);
    console.log(res.length === 0 ? 0 :Math.max(...res));

}

solution();